### PR TITLE
Refactor CharSetSolver and friends to avoid unbounded caching

### LINKED ISSUE #65417: Test failure System.Text.RegularExpressions.Tests.RegexKnownPatternTests.PatternsDataSet_ConstructRegexForAll_NonBacktracking
Run: [runtime-libraries-coreclr outerloop 20220215.3](https://dev.azure.com/dnceng/public/_build/results?buildId=1613576&view=ms.vss-test-web.build-test-results-tab&runId=44921124&paneView=debug&resultId=104290)

Failed test:
```
net7.0-windows-Release-x86-CoreCLR_release-Windows.Amd64.Server2022.Open

- System.Text.RegularExpressions.Tests.RegexKnownPatternTests.PatternsDataSet_ConstructRegexForAll_NonBacktracking

```
**Error message:**
```
System.OutOfMemoryException : Exception of type 'System.OutOfMemoryException' was thrown.


Stack trace
   at System.Collections.Concurrent.ConcurrentDictionary`2.GrowTable(Tables tables) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 1983
   at System.Collections.Concurrent.ConcurrentDictionary`2.TryAddInternal(TKey key, Nullable`1 nullableHashcode, TValue value, Boolean updateIfExists, Boolean acquireLock, TValue& resultingValue) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 993
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 1177
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.ApplyBinaryOp(BoolOp op, BDD a, BDD b) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 0
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.<>c.<ApplyBinaryOp>b__8_0(ValueTuple`3 key, BDDAlgebra algebra) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 176
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 1177
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.ApplyBinaryOp(BoolOp op, BDD a, BDD b) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 0
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.<>c.<ApplyBinaryOp>b__8_0(ValueTuple`3 key, BDDAlgebra algebra) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 176
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 1177
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.ApplyBinaryOp(BoolOp op, BDD a, BDD b) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 0
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.<>c.<ApplyBinaryOp>b__8_0(ValueTuple`3 key, BDDAlgebra algebra) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 176
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 1177
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.ApplyBinaryOp(BoolOp op, BDD a, BDD b) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 0
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.<>c.<ApplyBinaryOp>b__8_0(ValueTuple`3 key, BDDAlgebra algebra) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 176
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument) in /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:line 1177
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.ApplyBinaryOp(BoolOp op, BDD a, BDD b) in /_/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs:line 0
   at System.Text.RegularExpressions.Symbolic.BDDAlgebra.<>c.<ApplyBinaryOp>b__8_0(ValueTuple`3 key, BDDAlgebra algebra) in /_/src/libraries/System.Text.RegularExp
```

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/65417

CharSetSolver today is a singleton and holds on to a thread-safe cache that is augmented with every BDD produced by every regex processed.  Over time, then, the cache grows unbounded and represents a significant leak.  This change makes CharSetSolver no longer a singleton, with every regex creating its own CharSetSolver, with caches no longer shared across instances.  In doing so, we also remove the need for CharSetSolver's cache to be thread-safe, making it cheaper with a `Dictionary<>` instead of a `ConcurrentDictionary<>`.

The https://github.com/dotnet/runtime/blob/7e408a0fb69487a9113fab16be96fc3ea75a306d/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.KnownPattern.Tests.cs#L1430 test on my machine was taking ~90s prior to this change and takes ~35s after this change.

As part of this refactoring, other code has been refactored and cleaned up:
- CharSetSolver was the only type deriving from the abstract BDDAlgebra.  The base type has been removed and its functionality merged up into CharSetSolver.
- ICharAlgebra was the only type inheriting IBooleanAlgebra.  The latter has been removed and its APIs moved up into ICharAlgebra, which has also been renamed to ISolver.
- Similarly, UInt64Algebra and BitVectorAlgebra have been renamed to UInt64Solver and BitVectorSolver, respectively.
- Several files were renamed to match the types they contain.
- Several functions were parameterized with an integer value but the same value was passed in from every call site; that value has been moved to be a const in the method and the parameter removed.
- CharSetSolver had a complicated PrettyPrint method.  Most of it has been deleted in favor of using the set description functionality already present in RegexCharClass.
- Generic type parameter names for the implementation of sets (BDD, ulong, BitVector) were all over the map.  I've standardized on TSet.
- Several members of ISolver (formerly ICharAlgebra/IBooleanAlgebra) have been renamed: True is now Full, False is now Empty, and CharConstraint is now CreateFromChar.  IsSatisfiable and AreEquivalent were removed in favor of adding IsEmpty and IsFull members (AreEquivalent was only ever used with Full, and IsSatisfiable calls were replaced with !IsEmpty calls).
- Use of "predicate" has largely been removed.
- Replaced a few recursive algorithms with loops.
- Added some more comments.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDD.cs [hunk 1]
```diff
@@ -160,20 +135,18 @@ public ulong GetMin()
...
-        /// This equality is used in the _bddCache lookup.
-        /// </summary>
-        public override bool Equals(object? obj) =>
-            obj is BDD bdd &&
+        /// <summary>A shallow equality check that holds if ordinals are identical and one's are identical and zero's are identical.</summary>
+        public override bool Equals(object? obj) => Equals(obj as BDD);
+
+        /// <summary>A shallow equality check that holds if ordinals are identical and one's are identical and zero's are identical.</summary>
+        public bool Equals(BDD? bdd) =>
+            bdd is not null &&
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDD.cs:149]: NIT: I believe we don't have a test where we pass null, might it be worth adding a quick unittest for that?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs [hunk 2]
```diff
@@ -22,26 +22,26 @@ internal sealed class BDDRangeConverter
         /// Convert the set into an equivalent array of ranges.
         /// The ranges are nonoverlapping and ordered.
         /// </summary>
-        public static (uint, uint)[] ToRanges(BDD set, int maxBit)
+        public static (uint, uint)[] ToRanges(BDD set)
         {
-            Debug.Assert(0 <= maxBit && maxBit <= 31, "maxBit must be between 0 and 31");
+            const int MaxBit = 15;
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs:27]: NIT: Should we add a comment here why 15 was chosen?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs [hunk 3]
```diff
@@ -22,26 +22,26 @@ internal sealed class BDDRangeConverter
...
 
             if (set.IsEmpty)
                 return Array.Empty<(uint, uint)>();
 
             if (set.IsFull)
-                return new[] { (0u, ((uint)1 << maxBit << 1) - 1) }; //note: maxBit could be 31
+                return new[] { (0u, ((uint)1 << MaxBit << 1) - 1) }; //note: maxBit could be 31
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs:33]: this comment is no longer accurate

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs [hunk 4]
```diff
@@ -28,73 +28,73 @@ public BitVectorAlgebra(BDD[] minterms)
...
+            Full = BitVector.CreateTrue(minterms.Length);
         }
 
-        public BitVector False { get; }
-        public BitVector True { get; }
+        public BitVector Empty { get; }
+        public BitVector Full { get; }
 
-        public bool AreEquivalent(BitVector predicate1, BitVector predicate2) => predicate1.Equals(predicate2);
+        public bool IsFull(BitVector set) => set.Equals(Full);
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs:38]: NIT: this doesn't seem to be covered in our tests

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharSetSolver.cs [hunk 5]
```diff
@@ -2,327 +2,407 @@
...
-                CreateBDDFromChar(c);
-        }
-
-        private BDD CreateBDDFromChar(ushort c)
+        /// <summary>BDD for each ASCII character that returns true for that one character.</summary>
+        /// <remarks>This cache is shared amongst all CharSetSolver instances and is accessed in a thread-safe manner.</remarks>
+        private static readonly BDD?[] s_asciiCache = new BDD[128];
+        /// <summary>BDD that returns true for every non-ASCII character.</summary>
+        /// <remarks>This instance is shared amongst all CharSetSolver instances and is accessed in a thread-safe manner.</remarks>
+        private static BDD? s_nonAscii;
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharSetSolver.cs:23]: Does our existing test coverage guarantee that these will be accessed in a thread-safe manner? Or asked differently, should we expect something to break in our tests of the way these are accessed is ever changed incorrectly?

Comment (joperezr) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/CharSetSolver.cs:23]: oh NVM, I see that we just use Interlocked.CompareExchange.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs:38]: There's a non-trivial amount of code not covered by our tests, unfortunately.  I'd asked @olsaarik a few weeks back if he could take a pass at adding patterns/inputs that would cover more things, and I think he was planning to do so after he's done with the more pressing fixes we've discussed. Olli, is that still something you're planning to do?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs]: Deleted

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BDDRangeConverter.cs]: Done

**Review by joperezr — APPROVED** [review]
Not very familiar with this engine just yet, but changes look good to me.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs (hunk 4 above)
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs:38]: That had fallen off my mind, but yes, that is something I should do. I can take it up after the matching logic changes.

**Comment by olsaarik** [comment]
I read through the changes and it all looks good to me. This simplifies the BDD algebra code a lot :)

**Review by olsaarik — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVectorSolver.cs:38]: Thanks, Olli.  I'll open an issue to track that.

**Comment by stephentoub** [comment]
Thanks for reviewing, @olsaarik and @joperezr.

**Comment by AndyAyersMS** [comment]
Also regressions seen here: https://github.com/dotnet/perf-autofiling-issues/issues/4626
newplot - 2022-04-28T125938 962

**Comment by stephentoub** [comment]
Thanks, yeah, regression in time to construct the Regex instance is expected:
https://github.com/dotnet/runtime/issues/67915#issuecomment-1096979972
The change removed erroneous caching that lead to a potentially large leak over time but that also resulted in a benchmark just constructing the same Regex over and over being faster than it should have been (essentially it was caching all the interesting stuff for a regex from one construction to the next).

**Pull request merged**